src/basic/terminal-util.c

```diff
@@ -144,33 +144,48 @@ int read_one_char(FILE *f, char *ret, usec_t t, bool *need_nl) {
         return 0;
 }
 
-int ask_char(char *ret, const char *replies, const char *text, ...) {
+#define DEFAULT_ASK_REFRESH_USEC (2*USEC_PER_SEC)
+
+int ask_char(char *ret, const char *replies, const char *fmt, ...) {
+        _cleanup_free_ char *r_text = NULL;
+        _cleanup_free_ char *text = NULL;
+        va_list ap;
         int r;
 
         assert(ret);
         assert(replies);
-        assert(text);
+        assert(fmt);
+
+        va_start(ap, fmt);
+        r = vasprintf(&text, fmt, ap);
+        va_end(ap);
+        if (r < 0)
+                return -ENOMEM;
+
+        r_text = strjoin("\r", text);
+        if (!r_text)
+                return -ENOMEM;
 
         for (;;) {
-                va_list ap;
                 char c;
                 bool need_nl = true;
 
                 if (colors_enabled())
                         fputs(ANSI_HIGHLIGHT, stdout);
 
-                va_start(ap, text);
-                vprintf(text, ap);
-                va_end(ap);
+                fputs(r_text, stdout);
```

Comment:
uh oh, instead of generating text and then r_text I'd just use fputc('\r') first, and then vprintf(). This is in no way performance sensitive, and I much prefer simplicity over performance...